docs/data/data-grid/recipes-editing/BulkEditingNoSnap.js

```diff
@@ -0,0 +1,185 @@
+                  _action: 'delete',
+                };
+                if (!unsavedChangesRef.current.rowsBeforeChange[id]) {
+                  unsavedChangesRef.current.rowsBeforeChange[id] = row;
+                }
+                setHasUnsavedRows(true);
+                apiRef.current.updateRows([row]); // to trigger row render
+              }}
+            />,
+          ];
+        },
+      },
+      ...data.columns,
+    ];
+  }, [data.columns, unsavedChangesRef, apiRef]);
+
+  const processRowUpdate = (newRow, oldRow) => {
+    const rowId = newRow.id;
+
+    unsavedChangesRef.current.unsavedRows[rowId] = newRow;
+    if (!unsavedChangesRef.current.rowsBeforeChange[rowId]) {
+      unsavedChangesRef.current.rowsBeforeChange[rowId] = oldRow;
+    }
+    setHasUnsavedRows(true);
+    return newRow;
+  };
+
+  const discardChanges = () => {
+    setHasUnsavedRows(false);
+    apiRef.current.updateRows(
+      Object.values(unsavedChangesRef.current.rowsBeforeChange),
+    );
+    unsavedChangesRef.current = {
+      unsavedRows: {},
+      rowsBeforeChange: {},
+    };
+  };
+
+  const saveChanges = async () => {
+    try {
+      // Persist updates in the database
+      setIsSaving(true);
+      await new Promise((resolve) => {
+        setTimeout(resolve, 1000);
+      });
+
+      setIsSaving(false);
+      const rowsToDelete = Object.values(
+        unsavedChangesRef.current.unsavedRows,
+      ).filter((row) => row._action === 'delete');
+      if (rowsToDelete.length > 0) {
+        apiRef.current.updateRows(rowsToDelete);
+      }
+
+      setHasUnsavedRows(false);
+      unsavedChangesRef.current = {
+        unsavedRows: {},
+        rowsBeforeChange: {},
+      };
+    } catch (error) {
+      setIsSaving(false);
+    }
+  };
+
+  return (
+    <div style={{ width: '100%' }}>
+      <div style={{ marginBottom: 8 }}>
+        <LoadingButton
+          disabled={!hasUnsavedRows}
+          loading={isSaving}
+          onClick={saveChanges}
+          startIcon={<SaveIcon />}
+          loadingPosition="start"
+        >
+          <span>Save</span>
+        </LoadingButton>
+        <Button
+          disabled={!hasUnsavedRows || isSaving}
+          onClick={discardChanges}
+          startIcon={<RestoreIcon />}
+        >
+          Discard all changes
+        </Button>
+      </div>
+      <div style={{ height: 400 }}>
+        <DataGridPremium
+          {...data}
+          columns={columns}
+          apiRef={apiRef}
+          disableRowSelectionOnClick
+          unstable_cellSelection
+          processRowUpdate={processRowUpdate}
+          experimentalFeatures={{ clipboardPaste: true }}
+          unstable_ignoreValueFormatterDuringExport
+          sx={{
+            '& .MuiDataGrid-row.row--removed': {
+              backgroundColor: 'rgba(255, 170, 170, 0.3)',
+            },
+            '& .MuiDataGrid-row.row--edited': {
+              backgroundColor: 'rgba(255, 254, 176, 0.3)',
```

Comment:
Optional design enhancement: Would it look cooler to have more related colors for dark and light themes? Something like [this](https://github.com/mui/mui-x/blob/0c8e06c6ed571b4de4b12427a86610a20b55f18a/packages/grid/x-data-grid/src/components/containers/GridRootStyles.ts#L8-L16).

Reply:
Sure, I've changed the color a bit for dark mode.
Or did you mean something else?